File: benchmarks/throughput.py
#!/usr/bin/env python3
"""Throughput benchmark: measures MB/s for large output (seq 1 500000)"""

import asyncio
import json
import time
import sys

try:
    import websockets
except ImportError:
    print("Installing websockets...")
    import subprocess

    subprocess.check_call([sys.executable, "-m", "pip", "install", "websockets", "-q"])
    import websockets

HOST = "localhost"
PORT = 7682


async def run_benchmark():
    uri = f"ws://{HOST}:{PORT}/ws"
    total_bytes = 0
    start_time = None
    end_time = None
    prompt_count = 0

    async with websockets.connect(
        uri, subprotocols=["tty"], max_size=10 * 1024 * 1024
    ) as ws:
        # Send init
        init = json.dumps({"AuthToken": "", "columns": 80, "rows": 24})
        await ws.send(init.encode())

        # Wait for initial prompt
        async def wait_for_prompt():
            nonlocal prompt_count
            while True:
                msg = await ws.recv()
                if isinstance(msg, bytes) and len(msg) > 1 and msg[0] == 0x30:
                    text = msg[1:].decode("utf-8", errors="replace")
                    if "$" in text or "%" in text or "#" in text:
                        prompt_count += 1
                        if prompt_count >= 1:
                            break

        try:
            await asyncio.wait_for(wait_for_prompt(), timeout=10)
        except asyncio.TimeoutError:
            pass

        # Send throughput command
        cmd = "yes | head -c 3000000\r"
        payload = bytes([0x30]) + cmd.encode()
        start_time = time.perf_counter()
        await ws.send(payload)

        # Collect all output
        async def collect_output():
            nonlocal total_bytes, end_time
            idle_count = 0
            while True:
                try:
                    msg = await asyncio.wait_for(ws.recv(), timeout=3)
                    idle_count = 0
                except asyncio.TimeoutError:
                    idle_count += 1
                    if idle_count >= 2 and total_bytes > 50_000:
                        # No data for 6 seconds and we have some data
                        end_time = time.perf_counter()
                        return
                    continue

                if isinstance(msg, bytes) and len(msg) > 1 and msg[0] == 0x30:
                    data = msg[1:]
                    total_bytes += len(data)
                    text = data.decode("utf-8", errors="replace")
                    # Detect prompt return (command finished)
                    lines = text.split("\n")
                    for line in lines:
                        line = line.strip()
                        if (
                            line.endswith("$")
                            or line.endswith("%")
                            or line.endswith("#")
                        ):
                            if total_bytes > 50_000:
                                end_time = time.perf_counter()
                                return

        try:
            await asyncio.wait_for(collect_output(), timeout=60)
        except asyncio.TimeoutError:
            end_time = time.perf_counter()

    if start_time and end_time and total_bytes > 50_000:
        elapsed = end_time - start_time
        kbs = (total_bytes / 1024) / elapsed
        result = {
            "test": "throughput",
            "total_bytes": total_bytes,
            "elapsed_seconds": round(elapsed, 3),
            "throughput_kbs": round(kbs, 2),
        }
        print(json.dumps(result, indent=2))
    else:
        print(
            json.dumps({"error": "timeout or incomplete", "total_bytes": total_bytes})
        )


if __name__ == "__main__":
    asyncio.run(run_benchmark())

File: benchmarks/throughput.js
#!/usr/bin/env node
/**
 * Throughput benchmark: measures MB/s for large output (seq 1 500000)
 * Uses the globally available 'ws' npm package.
 */
'use strict';

const WebSocket = require('ws');

const HOST = process.env.HOST || 'localhost';
const PORT = process.env.BENCH_PORT || process.env.PORT || '7682';
const URI = `ws://${HOST}:${PORT}/ws`;

function run() {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(URI, ['tty']);
    ws.binaryType = 'nodebuffer';

    let totalBytes = 0;
    let startTime = null;
    let endTime = null;
    let promptReceived = false;
    let done = false;

    const timeout = setTimeout(() => {
      if (!done) {
        done = true;
        ws.terminate();
        resolve({ error: 'timeout' });
      }
    }, 90000);

    ws.on('error', (err) => {
      if (!done) { done = true; clearTimeout(timeout); ws.terminate(); reject(err); }
    });

    const send = (text) => {
      const payload = Buffer.concat([Buffer.from([0x30]), Buffer.from(text)]);
      ws.send(payload);
    };

    ws.on('open', () => {
      const init = JSON.stringify({ AuthToken: '', columns: 80, rows: 24 });
      ws.send(Buffer.from(init));
      setTimeout(() => send('echo __BENCH_READY__\r'), 2000);
    });

    ws.on('message', (data) => {
      if (done) return;
      if (!Buffer.isBuffer(data) || data.length < 2 || data[0] !== 0x30) return;

      const text = data.slice(1).toString('utf8');

      if (!promptReceived) {
        if (text.includes('__BENCH_READY__')) {
          promptReceived = true;
          startTime = process.hrtime.bigint();
          send('seq 1 500000\r');
        }
        return;
      }

      totalBytes += data.length - 1;

      if (text.includes('500000') && totalBytes > 2_000_000) {
        endTime = process.hrtime.bigint();
        if (!done) {
          done = true;
          clearTimeout(timeout);
          ws.terminate();

          const elapsedSec = Number(endTime - startTime) / 1e9;
          const mbs = (totalBytes / 1024 / 1024) / elapsedSec;
          resolve({
            test: 'throughput',
            total_bytes: totalBytes,
            elapsed_seconds: Math.round(elapsedSec * 1000) / 1000,
            throughput_mbs: Math.round(mbs * 100) / 100,
          });
        }
      }
    });
  });
}

run()
  .then((result) => {
    console.log(JSON.stringify(result, null, 2));
    process.exit(0);
  })
  .catch((err) => {
    console.error(JSON.stringify({ error: String(err) }));
    process.exit(1);
  });
